smarts/core/road_map.py

```diff
@@ -0,0 +1,437 @@
+        max_to_gen: int = 1,
+    ) -> List[RoadMap.Route]:
+        """ Routes will be returned in order of increasing road length """
+        # May be costly.
+        raise NotImplementedError()
+
+    def random_route(self, max_route_len: int = 10) -> RoadMap.Route:
+        raise NotImplementedError()
+
+    def waypoint_paths(
+        self,
+        pose: Pose,
+        lookahead: int,
+        within_radius: float = 5,
+        route: RoadMap.Route = None,
+    ) -> List[List[Waypoint]]:
+        """Computes equally-spaced Waypoints for all lane paths
+        up to lookahead waypoints ahead, starting on the Road containing
+        the nearest Lane aligned with the vehicle's pose within within_radius meters.
+        Constrains paths to the supplied route if specified."""
+        raise NotImplementedError()
+
+    class Lane:
+        @property
+        def lane_id(self) -> str:
+            raise NotImplementedError()
+
+        @property
+        def road(self) -> RoadMap.Road:
+            raise NotImplementedError()
+
+        @property
+        def speed_limit(self) -> float:
+            raise NotImplementedError()
+
+        @property
+        def length(self) -> float:
+            raise NotImplementedError()
+
+        @property
+        def in_junction(self) -> bool:
+            raise NotImplementedError()
+
+        @property
+        def index(self) -> int:
+            """ 0 is outer / right-most (relative to lane heading) lane on road. """
+            raise NotImplementedError()
+
+        @property
+        def lane_to_left(self) -> Tuple[RoadMap.Lane, bool]:
+            """Note: left is defined as 90 degrees clockwise relative to the lane heading.
+            Second result is True if lane is in the same direction as this one
+            May return None for lanes in junctions."""
+            raise NotImplementedError()
+
+        @property
+        def lane_to_right(self) -> Tuple[RoadMap.Lane, bool]:
+            """Note: right is defined as 90 degrees counter-clockwise relative to the lane heading.
+            Second result is True if lane is in the same direction as this one.
+            May return None for lanes in junctions."""
+            raise NotImplementedError()
+
+        @property
+        def incoming_lanes(self) -> List[RoadMap.Lane]:
+            raise NotImplementedError()
+
+        @property
+        def outgoing_lanes(self) -> List[RoadMap.Lane]:
+            raise NotImplementedError()
+
+        @property
+        def oncoming_lanes(self) -> List[RoadMap.Lane]:
+            raise NotImplementedError()
+
+        @property
+        def foes(self) -> Tuple[List[RoadMap.Lane], bool]:
+            """All lanes that in some way intersect with (cross) this one,
+            including those that have the same outgoing lane as this one,
+            and so might require right-of-way rules.
+            If right-of-way rules are known, these will be in order of priority,
```

Comment:
I think this is a detail for the implementation to work out.  I actually think the simplest thing to do here is to remove the "these will be in order of priority" comment and leave it up to the implementation completely.  (I'll probably do that.)

But this is also related to the "overlapping roads" question you asked above.  It's a tangled nest.

Reply:
Sorry, what I mean here is that for this method to be useful for priority in current form it would also have to return `self` or something other than just lanes. Otherwise there would be no point in mentioning priority at all.

Reply:
ah, now I get your concern. yeah, the ordering thing doesn't make much sense now for multiple reasons.  I'm going to remove the comment about it.